src/System.Private.CoreLib/shared/System/Threading/WaitHandle.cs

```diff
@@ -118,46 +125,58 @@ private bool WaitOneCore(int millisecondsTimeout, bool interruptible = true)
+                int waitResult;
+
+                SynchronizationContext context = SynchronizationContext.Current;
+                if (context != null && context.IsWaitNotificationRequired())
+                {
+                    var handles = new[] { waitHandle.DangerousGetHandle() };
+                    waitResult = context.Wait(handles, false, millisecondsTimeout);
+                    GC.KeepAlive(handles);
+                }
+                else
+                {
+                    waitResult = WaitOneCore(waitHandle.DangerousGetHandle(), millisecondsTimeout);
+                }
+
+                if (waitResult == WaitAbandoned)
+                {
+                    throw new AbandonedMutexException();
+                }
+
+                return waitResult != WaitTimeout;
             }
             finally
             {
-                waitHandle.DangerousRelease();
+                if (success)
+                    waitHandle.DangerousRelease();
             }
         }
 
-        public virtual bool WaitOne(TimeSpan timeout) => WaitOneCore(ToTimeoutMilliseconds(timeout));
-        public virtual bool WaitOne() => WaitOneCore(Timeout.Infinite);
-
-        public virtual bool WaitOne(int millisecondsTimeout, bool exitContext) => WaitOne(millisecondsTimeout);
-        public virtual bool WaitOne(TimeSpan timeout, bool exitContext) => WaitOne(timeout);
-
         /// <summary>
         /// Obtains all of the corresponding safe wait handles and adds a ref to each. Since the <see cref="SafeWaitHandle"/>
         /// property is publically modifiable, this makes sure that we add and release refs one the same set of safe wait
         /// handles to keep them alive during a multi-wait operation.
         /// </summary>
-        private static SafeWaitHandle[] ObtainSafeWaitHandles(
-            RuntimeThread currentThread,
-            WaitHandle[] waitHandles,
-            int numWaitHandles,
-            out SafeWaitHandle[] rentedSafeWaitHandles)
+        private static void ObtainSafeWaitHandles(
+            Span<WaitHandle> waitHandles,
+            out SafeWaitHandle[] safeWaitHandles,
+            out IntPtr[] unsafeWaitHandles)
         {
-            Debug.Assert(currentThread == RuntimeThread.CurrentThread);
             Debug.Assert(waitHandles != null);
-            Debug.Assert(numWaitHandles > 0);
-            Debug.Assert(numWaitHandles <= MaxWaitHandles);
-            Debug.Assert(numWaitHandles <= waitHandles.Length);
+            Debug.Assert(waitHandles.Length > 0);
+            Debug.Assert(waitHandles.Length <= MaxWaitHandles);
 
-            rentedSafeWaitHandles = currentThread.RentWaitedSafeWaitHandleArray(numWaitHandles);
-            SafeWaitHandle[] safeWaitHandles = rentedSafeWaitHandles ?? new SafeWaitHandle[numWaitHandles];
+            safeWaitHandles = new SafeWaitHandle[waitHandles.Length];
+            unsafeWaitHandles = new IntPtr[waitHandles.Length];
```

Comment:
I dropped the WaitHandle array renting when porting to CoreCLR. There are several approaches that could be used to bring it back:

1) Move the `Rent`/`Return` methods to shared `WaitHandle` and store the `WaitHandleArray` as a TLS variable.
2) Use `ArrayPool<T>.Shared` instead
3) Just allocate here, possibly switch `unsafeWaitHandles` to `stackalloc`.

Thoughts?

Reply:
@kouvel looked into this originally and ended up with the renting scheme. It was before we had Span.

Is the size of these arrays limited to 64 elements? If it is, I think it would be fine to allocate everything on stack.

Ultimately, this is a performance question. Pick the solution with the performance characteristics.

Reply:
It is limited to 64 elements, but I cannot `stackalloc` the first array unfortunately since it's array of reference types. (It doesn't come up often, but this is the second time in last two weeks when it would have been handy, for both performance and correctness in out-of-memory scenarios.)

Reply:
You can "stackalloc" it using a hack like this:

```csharp
struct Block8<T>
{
    T _1, _2, _3, _4, _5, _6, _7, _8;
}

struct Block64<T>
{
    Block8<T> _1, _2, _3, _4, _5, _6, _7, _8;
}

void foo()
{
    Block64<string> block;
    Span<string> span = new Span<string>(ref Unsafe.As<Block64<string>, string>(ref block), 64);

    ...
}
```

Note that this hack is not "free" - you still pay for the zero-initialization.

Reply:
I kinda figured the `Unsafe.As` part of the trick meanwhile, but it felt too wild to even suggest it. I'll try to benchmark it, just for the kicks.

Reply:
Hmm, no matter what I tried I just couldn't get your code to compile in any form.

I did some benchmarks and it looks like any sort of caching would have a measurable impact. I tried to simulate the effect of couple of long waits where garbage collection would force the arrays to get to a higher generation (by keeping last 1000 arrays referenced and then forcing `GC.Collect` and releasing the references; doing that in outer loop in cycles). The results showed that the impact could be measured and on the allocation / garbage collection in tight loop it could reach around 20% overhead compared to `ArrayPool.Shared.Rent/Return`. A custom `[ThreadStatic]` cache cut down another 14%, nearly on par with `stackalloc` (with `IntPtr` since I couldn't get it work with any trick for reference types). Then again, in big picture the differences are probably quite meaningless as long as any caching is used to reduce GC pressure.


|     Method |     Mean |     Error |    StdDev |
|----------- |---------:|----------:|----------:|
|  HeapAlloc | 39.21 ms | 0.7840 ms | 1.8784 ms |
|  ArrayPool | 31.75 ms | 0.6132 ms | 0.9178 ms |
|        Tls | 27.41 ms | 0.5692 ms | 0.8693 ms |
| StackAlloc | 27.01 ms | 0.5329 ms | 0.8137 ms |

Reply:
I ended up with a simple implementation with two `[ThreadStatic]` arrays that always have the `MaxWaitHandles` size. That's roughly 1Kb of allocated space per thread. It didn't seem worth it to bring over the whole `WaitHandleArray` logic to shared code part.

Alternatives are the following:
1) Move `WaitHandleArray` to shared code. It starts with smaller arrays that are gradually enlarged.
2) Limit the thread static arrays to some smaller size (eg. 4 handles would likely cover most wait scenarios in CoreCLR/CoreFX) and fallback to array allocation for anything bigger.
3) Use `ArrayPool.Shared` and let it do its magic. It has slightly bigger overhead, but it can allocate more than one array in case of reentrant waits and the real-world performance difference would likely be quite negligible.